Koc

Stepka opened the pull request:
Updated KOC

paulsj-80 commented:
- NOTE: don't take this all too seriously, some are just "what if"s,
  and some are probably out of scope for this particular pull request
  - haven't checked algorithm complexity and feasibility
  - split into several cases
    - zero test case should be extracted and put as the first one in
      source file
    - this promotes addition of further test cases and helps to
      diagnose problems faster
  - interface: what is the reasoning behind parameter split between
    constructor and factory, see this fragment:
    metric::KOC_factory<Record, metric::Grid4> simple_koc_factory(best_w_grid_size, best_h_grid_size, 0.5, 0.0, 300);    
	auto simple_koc = simple_koc_factory(dataset, num_clusters);
    - why not, for example:
      metric::KOC_factory<Record, metric::Grid4> simple_koc_factory;    
      auto simple_koc = simple_koc_factory(best_w_grid_size, best_h_grid_size, 0.5, 0.0, 300, dataset, num_clusters);
    - it should be clear in order to make class more usable
      - is it expected that factory allocates some expensive resources
        once, to be reused for subsequent calls? this doesn't happen in
        code
   - interface: is it normal use case to have different sigma for
     auto anomalies = simple_koc.check_if_anomaly(dataset, sigma);
     and
     auto assignments = simple_koc.result(dataset, sigma);
     - what behaviour is expected if sigma differs between those calls?
     - could / should it be included in simple_koc initializers if not?
   - interface: method naming; simple_koc.result() could describe more
     precisely what exactly is returned
     - variable is called assignments
     - zero-named clusters are checked
     - zero-named anomalies are checked
     - top_outlier() on contrary gives more precise understanding of
       what is returned (it is also a result of training)
     - ideally, function names reflect well understood and known concepts
       from respective field (such as anomaly, deviation, etc.), which
       the current code actually does, just this one detail
   - kohonen_outliers_clustering_tests.cpp:92
     - comment: anomalies or clusters?
   - is it correct to inherit KOC from SOM?
     - yes:
       - KOC seems to be type of SOM
       - interfaces are similar
     - no:
       - many methods are not virtual
       - if things are done differently in KOC, then it is not pure
         inheritance (child class should support all functionality of
         parent + some new functionality; is SOM that abstract?)
         - other way to check this is: are there any data members /
           functions in SOM which don't make sense with KOC? if "yes",
           then inheritance is not pure
           - for example, method encode() in SOM doesn't take sigma as a
             parameter; what if user calls KOC::train() without sigma?
             will the result be meaningful?
             - same about train()
           - what if KOC is cast to SOM, then encode() is called?
           - see Liskov Substitution Principle
       - possible solution would be to use composition instead of
         inheritance
  - KOC_factory constructors: why not use initializers for data members
    instead of assignment operator?
  - KOC.cpp:201: are there strong guarantees that min_cluster_size will
    exceed 1 with any input data? if no, there can be an endless loop

Stepka commented:
(1) haven't checked algorithm complexity and feasibility

	Not sure how can it be done, except Main test case

(2) split into several cases

	Done


(3) interface: what is the reasoning behind parameter split between constructor and factory
	
	The idea behind that split is the following: to hold structure of the KOC in the factory object (grid, metric, etc), and, otherwise, to create KOC object trained on specific dataset using () operator. Note: KOC is a model, but KOC_factory is a factory for a KOC. 
	
(4) interface: is it normal use case to have different sigma for check_if_anomaly(dataset, sigma) and result(dataset, sigma)
	
	Moved sigma to KOC constructor
	
(5) interface: method naming; 

	Renamed result() to assign_to_clusters(), top_outlier() to top_outliers(), and reduced return values for top_outliers() (to indexes and distances only)
	
(6) kohonen_outliers_clustering_tests.cpp:92 - comment: anomalies or clusters?

	Mixed anomalies and clusters. Anomalies is 0, clusters is starting from 1.
	
(7) is it correct to inherit KOC from SOM?

	Removed SOM inheritance and made som object as member of the KOC
	
(8) KOC_factory constructors: why not use initializers for data members instead of assignment operator?

	Done
	
(9) KOC.cpp:201: are there strong guarantees that min_cluster_size will exceed 1 with any input data? if no, there can be an endless loop

	Updated to use variable there. Reducing number of clusters for kmeans until 1 guarantees that.
